### PR TITLE
Build wrappers for HTTP objects

### PR DESCRIPTION
**Pull request by szymonlopaciuk**

### CODE REVIEW

File: inspire_mitmproxy/http/headers.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+            if key in header_dict:
+                header_dict[key].append(value)
+            else:
+                header_dict[key] = [value]
+
+        return cls(header_dict)
+
+    def to_dict(self) -> Dict[str, List[str]]:
+        return self.headers
```
Comment (david-caro) [inline inspire_mitmproxy/http/headers.py]: Return a copy of it instead.

File: inspire_mitmproxy/http/headers.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        try:
+            return self.headers[item.title()][0]
+        except KeyError:
+            return None
+
+    def __eq__(self, other) -> bool:
+        return self.headers == other.headers
+
+    def __repr__(self):
+        return repr(self.headers)
```
Comment (david-caro) [inline inspire_mitmproxy/http/headers.py]: `MITMHeaders(headers={"bla": ["blob", "bleb"]})`

File: inspire_mitmproxy/http/response.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from .utils import encoding_by_header
+
+
+class MITMResponse:
+    def __init__(
+        self,
+        status_code: int = 200,
+        status_message: Optional[str] = None,
+        body: Optional[str] = None,
+        headers: MITMHeaders = None,
```
Comment (david-caro) [inline inspire_mitmproxy/http/response.py]: should be `Optional[MTMHeaders]`

File: inspire_mitmproxy/http/headers.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                    (key.encode('ascii'), value.encode('ascii'))
+                )
+
+        return Headers(fields=fields)
+
+    def __getitem__(self, item: str) -> Optional[str]:
+        try:
+            return self.headers[item.title()][0]
+        except KeyError:
+            return None
```
Comment (david-caro) [inline inspire_mitmproxy/http/headers.py]:
```
headers['idontexist']
* KeyError
* value

headers.get('idontexist'[, default])
* value
* None/default
```

File: tests/integration/test_base_service.py [hunk 5]
```diff
@@ -65,75 +66,83 @@ def scenarios_dir(request):
...
+            method='GET',
+            url='https://host_b.local/api',
+            headers=MITMHeaders({
+                'Host': ['host_b.local'],
+                'Accept': ['application/json'],
+            })
+        )
+    )
+
+    assert response_service_2.body == 'test_scenario1/TestServiceB/0'
```
Comment (david-caro) [inline tests/integration/test_base_service.py]:
```python

set_config_req ..
get_config_hosta ...
get_config_hostb ...

response = dispatcher.process_request(set_config_req)
assert response.status_code == 201

response = dispatcher.process_request(get_config_hosta)
assert response.body = ...

response = dispatcher.process_request(get_config_hostb)
assert response.body = ...
```

File: tests/unit/test_http_response.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+def test_response_to_dict():
+    result = TEST_RESPONSE.to_dict()
+    expected = TEST_DICT_RESPONSE
+
+    assert result == expected
+
+
+def test_responses_from_bytes_and_str_equal():
+    assert MITMResponse.from_dict(TEST_DICT_RESPONSE) \
+        == MITMResponse.from_dict(TEST_DICT_RESPONSE_WITH_BYTES_BODY)
```
Comment (david-caro) [inline tests/unit/test_http_response.py]: result == expected

File: tests/unit/test_http_utils.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        'Value2',
+    ]
+})
+
+
+def test_encoding_by_header():
+    result = encoding_by_header(TEST_HEADERS)
+    expected = 'ASCII'
+
+    assert result == expected
```
Comment (david-caro) [inline tests/unit/test_http_utils.py]: missing 'utf-8' default value test

File: inspire_mitmproxy/http/response.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            http_version=response.http_version,
+        )
+
+    @classmethod
+    def from_dict(cls, response: Dict[str, Any]) -> 'MITMResponse':
+        encoding = encoding_by_header(MITMHeaders.from_dict(response['headers']))
+
+        body = response['body']
+        if isinstance(body, bytes):
+            body = body.decode(encoding or 'utf-8')
```
Comment (david-caro) [inline inspire_mitmproxy/http/response.py]: No need for the `or ...`

File: inspire_mitmproxy/http/response.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        headers: MITMHeaders = None,
+        original_encoding: Optional[str] = None,
+        http_version: Optional[str] = None,
+    ) -> None:
+        self.status_code = status_code
+        self.status_message = status_message or RESPONSES[status_code]
+        self.body = body or ''
+        self.headers = headers or MITMHeaders({})
+        self.original_encoding = original_encoding
+        self.http_version = http_version or 'HTTP/1.1'
```
Comment (david-caro) [inline inspire_mitmproxy/http/response.py]: Guess the encoding from the headers here if not passed, and not in the 'from_*' methods:
```
 +        encoding = encoding_by_header(MITMHeaders.from_mitmproxy(response.headers)) 
~``

File: inspire_mitmproxy/http/response.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            original_encoding=encoding
+        )
+
+    def to_mitmproxy(self) -> HTTPResponse:
+        return HTTPResponse(
+            http_version='HTTP/1.1',
+            status_code=self.status_code,
+            reason=self.status_message,
+            headers=self.headers.to_mitmproxy(),
+            content=self.body.encode(self.original_encoding or 'utf-8'),
```
Comment (david-caro) [inline inspire_mitmproxy/http/response.py]: If we always set the encoding in the init, we don't need the `or...`

File: tests/unit/test_http_utils.py (hunk 7 above)
Comment (szymonlopaciuk) [reply tests/unit/test_http_utils.py]: Done

File: inspire_mitmproxy/http.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        return (
+            self.url == other.url
+            and self.method == other.method
+            and self.body == other.body
+            and self.headers == other.headers
+        )
+
+    def __repr__(self):
+        return f'MITMRequest("{self.url}", "{self.method}", ' \
+            f'headers={self.headers}, body="{self.body}")'
```
Comment (david-caro) [inline inspire_mitmproxy/http.py]: Better use `repr`

File: inspire_mitmproxy/services/base_service.py [hunk 12]
```diff
@@ -85,4 +87,9 @@ def get_responses_for_active_scenario(self) -> List[dict]:
             if response.is_file() and response.suffix == '.yaml'
         ]
 
+        responses = [
+            (MITMRequest.from_dict(v['request']), MITMResponse.from_dict(v['response']))
+            for v in responses
+        ]
+
```
Comment (david-caro) [inline inspire_mitmproxy/services/base_service.py]:
```                                                                                                                            
    1         responses = [                                                           
    2             self._get_recorded_entry(response)                                  
    3             for response in responses_dir.iterdir()                             
    4             if response.is_file() and response.suffix == '.yaml'                
    5         ]                                                                       
    6                                                                                 
    7         return responses                                                        
    8                                                                                 
    9     @staticmethod                                                               
   10     def _get_recorded_entry(response_file):                                     
   11         text = response.read_text()                                             
   12         pair = yaml.load(text)                                                  
   13         return {                                                                
   14             'request': MITMRequest.from_dict(pair['request']),                  
   15             'response': MITMResponse.from_dict(pair['response']),               
   16         }                                                                       
   17                                                                                 
```

File: inspire_mitmproxy/services/management_service.py [hunk 13]
```diff
@@ -93,38 +93,35 @@ def get_scenarios(self) -> dict:
...
-            },
-            'body': body,
-            'headers': {
-                'Content-Type': ['application/json'],
+        return MITMResponse(
+            status_code=code,
+            body=body,
+            headers=MITMHeaders({
+                'Content-Type': ['application/json; encoding=UTF-8'],
                 'Server': ['inspire-mitmproxy/' + get_current_version(getcwd())]
```
Comment (david-caro) [inline inspire_mitmproxy/services/management_service.py]:
```
autosemver.packaging.get_current_version(
    project_name='myprojectname'
)
```

File: inspire_mitmproxy/services/whitelist_service.py [hunk 14]
```diff
@@ -22,8 +22,9 @@
 
 """Service which allows all requests outside"""
 
-from .errors import DoNotIntercept
-from .services import BaseService
+from inspire_mitmproxy.errors import DoNotIntercept
+from inspire_mitmproxy.http import MITMRequest
+from inspire_mitmproxy.services import BaseService
```
Comment (david-caro) [inline inspire_mitmproxy/services/whitelist_service.py]: Relative import

File: inspire_mitmproxy/dispatcher.py [hunk 15]
```diff
@@ -27,11 +27,12 @@
 
 from mitmproxy.http import HTTPFlow, HTTPResponse
 
+from inspire_mitmproxy.services.whitelist_service import WhitelistService
```
Comment (david-caro) [inline inspire_mitmproxy/dispatcher.py]: Relative import

File: tests/integration/test_management_service.py [hunk 16]
```diff
@@ -118,68 +121,74 @@ def test_management_service_get_scenarios(fake_scenarios_dir, dispatcher):
...
+    result = dispatcher.process_request(
+        MITMRequest(
+            method='GET',
+            url='http://mitm-manager.local/config',
+            headers=MITMHeaders({
+                'Host': ['mitm-manager.local'],
+                'Accept': ['application/json'],
+            })
+        )
+    )
```
Comment (david-caro) [inline tests/integration/test_management_service.py]: Put some makeup on it

Comment (david-caro) [reply tests/integration/test_management_service.py]: :lipstick:

**Pull request merged**